libkbfs/tlf_journal.go

```diff
@@ -0,0 +1,611 @@
+
+			case <-j.needShutdownCh:
+				j.log.CDebugf(ctx,
+					"Got shutdown signal for %s", j.tlfID)
+				needShutdown = true
+			}
+
+			errCh = nil
+			bwCancel()
+			bwCancel = nil
+			if needShutdown {
+				return
+			}
+
+		case bws == TLFJournalBackgroundWorkEnabled && errCh == nil:
+			if j.bwDelegate != nil {
+				j.bwDelegate.OnNewState(ctx, bwIdle)
+			}
+			// Idle state.
+			j.log.CDebugf(
+				ctx, "Waiting for the work signal for %s",
+				j.tlfID)
+			select {
+			case <-j.hasWorkCh:
+				j.log.CDebugf(
+					ctx, "Got work signal for %s", j.tlfID)
+				bwCtx, cancel := context.WithCancel(ctx)
+				errCh = j.doBackgroundWork(bwCtx)
+				bwCancel = cancel
+
+			case <-j.needPauseCh:
+				j.log.CDebugf(ctx,
+					"Got pause signal for %s", j.tlfID)
+				bws = TLFJournalBackgroundWorkPaused
+
+			case <-j.needShutdownCh:
+				j.log.CDebugf(ctx,
+					"Got shutdown signal for %s", j.tlfID)
+				return
+			}
+
+		case bws == TLFJournalBackgroundWorkPaused:
+			if j.bwDelegate != nil {
+				j.bwDelegate.OnNewState(ctx, bwPaused)
+			}
+			// Paused state.
+			j.log.CDebugf(
+				ctx, "Waiting to resume background work for %s",
+				j.tlfID)
+			select {
+			case <-j.needResumeCh:
+				j.log.CDebugf(ctx,
+					"Got resume signal for %s", j.tlfID)
+				bws = TLFJournalBackgroundWorkEnabled
+
+			case <-j.needShutdownCh:
+				j.log.CDebugf(ctx,
+					"Got shutdown signal for %s", j.tlfID)
+				return
+			}
+
+		default:
+			j.log.CErrorf(
+				ctx, "Unknown TLFJournalBackgroundStatus %s",
+				bws)
+			return
+		}
+	}
+}
+
+// doBackgroundWork currently only does auto-flushing. It assumes that
+// ctx is canceled when the background processing should stop.
+//
+// TODO: Handle garbage collection too.
+func (j *tlfJournal) doBackgroundWork(ctx context.Context) <-chan error {
+	errCh := make(chan error, 1)
+	go func() {
+		errCh <- j.flush(ctx)
+	}()
+	return errCh
+}
+
+// We don't guarantee that pause/resume requests will be processed in
+// strict FIFO order. In particular, multiple pause requests are
+// collapsed into one (also multiple resume requests), so it's
+// possible that a pause-resume-pause sequence will be processed as
+// pause-resume. But that's okay, since these are just for infrequent
+// ad-hoc testing.
+
+func (j *tlfJournal) pauseBackgroundWork() {
+	select {
+	case j.needPauseCh <- struct{}{}:
+	default:
+	}
+}
+
+func (j *tlfJournal) resumeBackgroundWork() {
+	select {
+	case j.needResumeCh <- struct{}{}:
+	default:
+	}
+}
+
+func (j *tlfJournal) flush(ctx context.Context) (err error) {
+	flushedBlockEntries := 0
+	flushedMDEntries := 0
+	defer func() {
+		if err != nil {
+			j.deferLog.CDebugf(ctx,
+				"Flushed %d block entries and %d MD entries "+
+					"for %s, but got error %v",
+				flushedBlockEntries, flushedMDEntries,
+				j.tlfID, err)
+		}
+	}()
+
+	// TODO: Interleave block flushes with their related MD
+	// flushes.
+
+	// TODO: Parallelize block puts.
+
+	for {
+		flushed, err := j.flushOneBlockOp(ctx)
```

Comment:
Pausing and shutting down will cause the worker goroutine to exit -- we pass in bwCtx to doBackgroundWork (which passes it to the worker goroutine), and we call the corresponding bwCancel function when we exit out of the busy state. There are some tests that exercise this, and without that, they would hang.

I added some comments to call this out a bit better.